modules/dag-l0/src/test/scala/org/tessellation/dag/l0/domain/snapshot/programs/SnapshotBinaryFeeCalculatorSuite.scala

```diff
@@ -0,0 +1,141 @@
+
+import eu.timepit.refined.types.numeric._
+import org.scalacheck.Gen
+import weaver.MutableIOSuite
+import weaver.scalacheck.Checkers
+
+object SnapshotBinaryFeeCalculatorSuite extends MutableIOSuite with Checkers {
+  type Res = (KryoSerializer[IO], Hasher[IO], SecurityProvider[IO])
+
+  override def sharedResource: Resource[IO, Res] =
+    for {
+      implicit0(ks: KryoSerializer[IO]) <- KryoSerializer.forAsync[IO](sharedKryoRegistrar)
+      sp <- SecurityProvider.forAsync[IO]
+      implicit0(j: JsonSerializer[IO]) <- JsonSerializer.forSync[IO].asResource
+      h = Hasher.forSync[IO](new HashSelect {
+        def select(ordinal: SnapshotOrdinal): HashLogic = JsonHash
+      })
+    } yield (ks, h, sp)
+
+  implicit val configShow: Show[FeeCalculatorConfig] = Show.show[FeeCalculatorConfig](_.toString)
+
+  test("minumum value returned when address unavailable") { res =>
+    implicit val (ks, h, _) = res
+
+    val gen = for {
+      config <- feeCalculatorConfigGen
+      binary <- signedOf(stateChannelSnapshotBinaryGen(zeroLongGen))
+    } yield (config, binary)
+
+    forall(gen) {
+      case (config, binary) =>
+        val calculator = SnapshotBinaryFeeCalculator.make(config, () => None)
+        val actual = calculator.fee(binary, GlobalSnapshotInfo.empty)
+        expect.same(Long.MinValue, actual)
+    }
+  }
+
+  test("minumum value returned when balance unavailable") { res =>
+    implicit val (ks, h, _) = res
+
+    val gen = for {
+      config <- feeCalculatorConfigGen
+      address <- addressGen
+      binary <- signedOf(stateChannelSnapshotBinaryGen(zeroLongGen))
+    } yield (config, address, binary)
+
+    forall(gen) {
+      case (config, address, binary) =>
+        val calculator = SnapshotBinaryFeeCalculator.make(config, () => address.some)
+        val actual = calculator.fee(binary, GlobalSnapshotInfo.empty)
+        expect.same(Long.MinValue, actual)
+    }
+  }
+
+  test("return computed value") { res =>
+    implicit val (ks, h, _) = res
+
+    val gen = for {
+      config <- feeCalculatorConfigGen
+      address <- addressGen
+      binary <- signedOf(stateChannelSnapshotBinaryGen(zeroLongGen))
+      balance <- balanceGen
+    } yield (config, address, binary, balance)
+
+    forall(gen) {
+      case (config, address, binary, balance) =>
+        val calculator = SnapshotBinaryFeeCalculator.make(config, () => address.some)
+        // NOTE: current calculation is as follows:
+        val kbyteSize = binary.value.content.length / 1024.0
+        val feePerKb: Long = (binary.fee.value.value / kbyteSize).toLong
+
+        val workAmount: Double = kbyteSize * config.computationalCost.value
+
+        val optionalTip = Math.max(0L, feePerKb - config.baseFee.value)
+
+        val workMultiplier = 1 / (1 + balance.value.value * config.stakingWeight.value)
+
+        val calculatedFee: Double = config.baseFee.value * workAmount * workMultiplier + optionalTip
+        val expected = calculatedFee.toLong
```

Comment:
I'm not sure that copying the same code into test actually creates a meaningful test 🤔 You may have an error in the algorithm and copying all the steps here "as is" won't catch that. If we want to mimic some algorithm, I'd opt for pre-computing manually output for few inputs and use it to compare the computed output from algorithm.

Reply:
changed